File: src/main/java/com/github/grishberg/cad3d/keyboard/ThumbConnections.java
package com.github.grishberg.cad3d.keyboard;

import static com.github.grishberg.cad3d.keyboard.KeyPlaceholder.placeHolderLeft;
import static com.github.grishberg.cad3d.keyboard.KeyPlaceholder.placeHolderRight;
import static com.github.grishberg.cad3d.keyboard.Utils.hull;
import static com.github.grishberg.cad3d.keyboard.Utils.union;

import com.github.grishberg.cad3d.keyboard.cfg.KeyboardConfig;
import eu.printingin3d.javascad.models.Abstract3dModel;
import java.util.ArrayList;

public class ThumbConnections {

    private KeyboardConfig cfg;
    private final KeyPlace keyPlace;

    private final ArrayList<Abstract3dModel> models = new ArrayList<>();


    public ThumbConnections(KeyboardConfig cfg, KeyPlace keyPlace) {
        this.cfg = cfg;
        this.keyPlace = keyPlace;
    }

    public Abstract3dModel buildThumbPlaceConnections() {
        models.clear();

        addHull(
            ThumbKeyPlace.place1(placeHolderLeft()),
            ThumbKeyPlace.place2(placeHolderRight())
        );

        addHull(
            ThumbKeyPlace.place2(placeHolderLeft()),
            ThumbKeyPlace.place3(placeHolderRight())
        );

        return union(models);
    }

    private void addHull(Abstract3dModel... children) {
        models.add(hull(children));
    }

}

File: cad3d/src/main/java/com/github/grishberg/cad3d/keyboard/ThumbConnections.kt
package com.github.grishberg.cad3d.keyboard

import com.github.grishberg.cad3d.keyboard.KeyPlaceholder.placeHolderLeft
import com.github.grishberg.cad3d.keyboard.KeyPlaceholder.placeHolderRight
import com.github.grishberg.cad3d.keyboard.cfg.KeyboardConfig
import com.github.grishberg.cad3d.plugin.cfg.ThumbClusterMode
import com.github.grishberg.javascad.models.Model

class ThumbConnections(private val cfg: KeyboardConfig, private val thumbKeyPlace: ThumbKeyPlace) {

    private val models = ArrayList<Model>()
    fun buildThumbPlaceConnections(): Model {
        models.clear()

        return when(cfg.thumbClusterSettings.type){
            ThumbClusterMode.SingleColumn3Buttons -> create3ThumbsConnection()
            ThumbClusterMode.SingleColumn4Buttons -> create3ThumbsConnection()
            ThumbClusterMode.TwoRows5Buttons -> createTwoRows5Buttons()
        }
    }

    private fun create3ThumbsConnection(): Model {
        addHull(
            thumbKeyPlace.placeR(placeHolderLeft()), thumbKeyPlace.placeM(placeHolderRight())
        )
        addHull(
            thumbKeyPlace.placeM(placeHolderLeft()), thumbKeyPlace.placeL(placeHolderRight())
        )
        return Utils.union(models)
    }

    private fun createTwoRows5Buttons(): Model {
        addHull(
            thumbKeyPlace.placeR(placeHolderLeft()), thumbKeyPlace.placeM(placeHolderRight())
        )
        addHull(
            thumbKeyPlace.placeM(placeHolderLeft()), thumbKeyPlace.placeL(placeHolderRight())
        )

        addHull(
            thumbKeyPlace.placeR2(placeHolderLeft()), thumbKeyPlace.placeL2(placeHolderRight())
        )
        return Utils.union(models)
    }

    private fun addHull(vararg children: Model) {
        models.add(Utils.hull(*children))
    }
}
